### PR TITLE
MYB-65 FavoritesDao

### PR DESCRIPTION
**Pull request by jcg22cabeza**

### CODE REVIEW

File: app/src/main/java/com/mybus/model/UsageTrackable.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.mybus.model;
+
+/**
+ * Created by Julian Gonzalez <jgonzalez@devspark.com>
+ */
+public interface UsageTrackable {
```
Comment (pablotdl) [inline app/src/main/java/com/mybus/model/UsageTrackable.java:7]: Javadoc

File: app/src/main/java/com/mybus/dao/RealmDao.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+    /**
+     * @param field name of the field
+     * @param value of the field
+     * @return a copy of a realObject item finding by a field
+     */
+    public T getByField(String field, Date value) {
+        return mRealm.copyFromRealm(mRealm.where(mType).equalTo(field, value).findFirst());
+    }
+}
```
Comment (pablotdl) [inline app/src/main/java/com/mybus/dao/RealmDao.java:192]: newline

File: app/src/main/java/com/mybus/dao/RealmDao.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    public RealmResults<T> getAll() {
+        return mRealm.where(mType).findAll();
+    }
+
+    /**
+     * Increment the usage count for a given item.
+     * @param id
+     */
+    public void updateItemUsageCount(Long id) {
+        if (!(UsageTrackable.class).isAssignableFrom(mType)) {
```
Comment (pablotdl) [inline app/src/main/java/com/mybus/dao/RealmDao.java]: the parenthesis in UsageTrackable is not needed

File: app/src/main/java/com/mybus/dao/RealmDao.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            return;
+        }
+        UsageTrackable item = (UsageTrackable) mRealm.where(mType).equalTo("id", id).findFirst();
```
Comment (pablotdl) [inline app/src/main/java/com/mybus/dao/RealmDao.java:83]: getById?

File: app/src/main/java/com/mybus/dao/RealmDao.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        }
+    }
+
+    /**
+     * Remove an item from realm
+     * @param id To find the item
+     * @return
+     */
+    public boolean remove(Long id) {
+        T item = mRealm.where(mType).equalTo("id", id).findFirst();
```
Comment (pablotdl) [inline app/src/main/java/com/mybus/dao/RealmDao.java:53]: getById?

File: app/src/main/java/com/mybus/dao/RealmDao.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        if (item != null) {
+            mRealm.beginTransaction();
+            item.deleteFromRealm();
+            mRealm.commitTransaction();
```
Comment (pablotdl) [inline app/src/main/java/com/mybus/dao/RealmDao.java]: doesn't this need  try/finally?

File: app/src/main/java/com/mybus/dao/RealmDao.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        if (item == null) {
+            return null;
+        }
+        try {
+            mRealm.beginTransaction();
+            T updated = mRealm.copyToRealmOrUpdate(item);
+            mRealm.commitTransaction();
+            return updated;
+        } catch (IllegalArgumentException e) {
+            return null;
```
Comment (pablotdl) [inline app/src/main/java/com/mybus/dao/RealmDao.java]: Why would you get an IllegalArgumentException here and why would you swallow it and return null

File: app/src/main/java/com/mybus/dao/RealmDao.java (hunk 4 above)
Comment (jcg22cabeza) [inline app/src/main/java/com/mybus/dao/RealmDao.java:83]: GetBy\* uses copyFromRealm:
Which makes a standalone in-memory copy of an already persisted {@link RealmObject}. This is a deep copy that will copy
     \* all referenced objects up to the defined depth.
     \* The copied object(s) are all detached from Realm so they will no longer be automatically updated. This means
     \* that the copied objects might contain data that are no longer consistent with other managed Realm objects.

When you need to update/remove a item you need an attached copy.

File: app/src/main/java/com/mybus/dao/RealmDao.java (hunk 5 above)
Comment (jcg22cabeza) [inline app/src/main/java/com/mybus/dao/RealmDao.java:53]: response above

File: app/src/main/java/com/mybus/dao/RealmDao.java [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        return mRealm.copyFromRealm(mRealm.where(mType).equalTo(field, value).findFirst());
+    }
+
+    /**
+     * @param field name of the field
+     * @param value of the field
+     * @return a copy of a realObject item finding by a field
+     */
+    public T getByField(String field, Date value) {
+        return mRealm.copyFromRealm(mRealm.where(mType).equalTo(field, value).findFirst());
```
Comment (pablotdl) [inline app/src/main/java/com/mybus/dao/RealmDao.java]: What if findFirst returns null?

File: app/src/main/java/com/mybus/dao/RealmDao.java (hunk 7 above)
Comment (jcg22cabeza) [inline app/src/main/java/com/mybus/dao/RealmDao.java]: copyToRealmOrUpdate() javadoc:

@throws java.lang.IllegalArgumentException if the object is {@code null} or doesn't have a Primary key defined
     \*  or it belongs to a Realm instance in a different thread.

File: app/src/main/java/com/mybus/dao/RealmDao.java (hunk 8 above)
Comment (ldimitroff) [inline app/src/main/java/com/mybus/dao/RealmDao.java]: You should create a method as copyFromRealm checking if findFirst returned null in order not to get the IllegalArgumentException, returning null as the getByField didn't match any value.
Also we can make an instrumentation test on this behavior.

**Comment by pablotdl** [comment]
Better.

**Pull request merged**